### ISSUE TITLE
pressing many mod-taps at the same time gives weird results

### ISSUE THREAD

**Issue by okke-formsma** [zmkfirmware/zmk#692]
In the following log, you see a weird behavior. Four mod-tap keys are pressed at the same time, but they do not activate after 150 ms. Instead, only one of the mods activates after 150 ms, and the others only activate for a short time when the first one was released.

I think this has to do with the position-state-changed events being queued at the same time, instead of that the first event is fully handled, then the next, etc.

```
[00:32:10.889,648] <dbg> zmk.kscan_gpio_read_0: Sending event at 1,1 state on
[00:32:10.889,648] <dbg> zmk.kscan_gpio_read_0: Sending event at 1,2 state on
[00:32:10.889,678] <dbg> zmk.kscan_gpio_read_0: Sending event at 1,3 state on
[00:32:10.889,678] <dbg> zmk.kscan_gpio_read_0: Sending event at 1,4 state on
[00:32:10.889,739] <dbg> zmk.zmk_kscan_process_msgq: Row: 1, col: 1, position: 10, pressed: true
[00:32:10.889,770] <dbg> zmk.position_state_changed_listener: 10 bubble (no undecided hold_tap active)
[00:32:10.889,770] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:32:10.889,801] <dbg> zmk.zmk_kscan_process_msgq: Row: 1, col: 2, position: 11, pressed: true
[00:32:10.889,831] <dbg> zmk.position_state_changed_listener: 11 bubble (no undecided hold_tap active)
[00:32:10.889,862] <dbg> zmk.zmk_keymap_apply_position_state: layer: 0 position: 10, binding name: MOD_TAP
[00:32:10.889,923] <dbg> zmk.on_hold_tap_binding_pressed: 10 new undecided hold_tap
[00:32:10.889,953] <dbg> zmk.position_state_changed_listener: 10 capturing 11 down event
[00:32:10.889,953] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:32:10.889,953] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:32:10.889,984] <dbg> zmk.zmk_kscan_process_msgq: Row: 1, col: 3, position: 12, pressed: true
[00:32:10.890,014] <dbg> zmk.position_state_changed_listener: 10 capturing 12 down event
[00:32:10.890,014] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:32:10.890,014] <dbg> zmk.zmk_kscan_process_msgq: Row: 1, col: 4, position: 13, pressed: true
[00:32:10.890,045] <dbg> zmk.position_state_changed_listener: 10 capturing 13 down event
[00:32:10.890,075] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:32:11.040,039] <dbg> zmk.decide_hold_tap: 10 decided hold (tap-preferred event 3)
[00:32:11.040,069] <dbg> zmk.on_keymap_binding_pressed: position 10 keycode 0x700e0
[00:32:11.040,100] <dbg> zmk.hid_listener_keycode_pressed: usage_page 0x07 keycode 0xe0 implicit_mods 0x00 explicit_mods 0x00
[00:32:11.040,100] <dbg> zmk.zmk_hid_register_mod: Modifier 0 count 1
[00:32:11.040,130] <dbg> zmk.zmk_hid_register_mod: Modifiers set to 0x01
[00:32:11.040,130] <dbg> zmk.zmk_hid_implicit_modifiers_press: Modifiers set to 0x01
[00:32:11.040,130] <dbg> zmk.zmk_endpoints_send_report: usage page 0x07
[00:32:11.040,191] <dbg> zmk.release_captured_events: Releasing key position event for position 11 pressed
[00:32:11.040,222] <dbg> zmk.position_state_changed_listener: 11 bubble (no undecided hold_tap active)
[00:32:11.040,252] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:32:11.040,252] <dbg> zmk.release_captured_events: Releasing key position event for position 12 pressed
[00:32:11.040,252] <dbg> zmk.position_state_changed_listener: 12 bubble (no undecided hold_tap active)
[00:32:11.040,283] <dbg> zmk.zmk_keymap_apply_position_state: layer: 0 position: 11, binding name: MOD_TAP
[00:32:11.040,344] <dbg> zmk.on_hold_tap_binding_pressed: 11 new undecided hold_tap
[00:32:11.040,344] <dbg> zmk.position_state_changed_listener: 11 capturing 12 down event
[00:32:11.040,374] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:32:11.040,374] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:32:11.050,506] <dbg> zmk.release_captured_events: Releasing key position event for position 13 pressed
[00:32:11.050,506] <dbg> zmk.position_state_changed_listener: 11 capturing 13 down event
[00:32:11.050,506] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:32:16.662,689] <dbg> zmk.kscan_gpio_read_0: Sending event at 1,4 state off
[00:32:16.662,719] <dbg> zmk.zmk_kscan_process_msgq: Row: 1, col: 4, position: 13, pressed: false
[00:32:16.662,750] <dbg> zmk.decide_hold_tap: 11 decided hold (tap-preferred event 3)
[00:32:16.662,780] <dbg> zmk.on_keymap_binding_pressed: position 11 keycode 0x700e3
[00:32:16.662,811] <dbg> zmk.hid_listener_keycode_pressed: usage_page 0x07 keycode 0xe3 implicit_mods 0x00 explicit_mods 0x00
[00:32:16.662,811] <dbg> zmk.zmk_hid_register_mod: Modifier 3 count 1
[00:32:16.662,811] <dbg> zmk.zmk_hid_register_mod: Modifiers set to 0x09
[00:32:16.662,841] <dbg> zmk.zmk_hid_implicit_modifiers_press: Modifiers set to 0x09
[00:32:16.662,841] <dbg> zmk.zmk_endpoints_send_report: usage page 0x07
[00:32:16.662,902] <dbg> zmk.release_captured_events: Releasing key position event for position 12 pressed
[00:32:16.662,933] <dbg> zmk.position_state_changed_listener: 12 bubble (no undecided hold_tap active)
[00:32:16.662,933] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:32:16.662,933] <dbg> zmk.release_captured_events: Releasing key position event for position 13 pressed
[00:32:16.662,963] <dbg> zmk.position_state_changed_listener: 13 bubble (no undecided hold_tap active)
[00:32:16.662,994] <dbg> zmk.zmk_keymap_apply_position_state: layer: 0 position: 12, binding name: MOD_TAP
[00:32:16.663,055] <dbg> zmk.on_hold_tap_binding_pressed: 12 new undecided hold_tap
[00:32:16.663,085] <dbg> zmk.position_state_changed_listener: 12 capturing 13 down event
[00:32:16.663,085] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:32:16.663,085] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:32:16.663,085] <dbg> zmk.position_state_changed_listener: 12 capturing 13 up event
[00:32:16.663,116] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:32:16.678,131] <dbg> zmk.kscan_gpio_read_0: Sending event at 1,1 state off
[00:32:16.678,131] <dbg> zmk.kscan_gpio_read_0: Sending event at 1,2 state off
[00:32:16.678,161] <dbg> zmk.kscan_gpio_read_0: Sending event at 1,3 state off
[00:32:16.678,222] <dbg> zmk.zmk_kscan_process_msgq: Row: 1, col: 1, position: 10, pressed: false
[00:32:16.678,253] <dbg> zmk.decide_hold_tap: 12 decided hold (tap-preferred event 3)
[00:32:16.678,283] <dbg> zmk.on_keymap_binding_pressed: position 12 keycode 0x700e2
[00:32:16.678,314] <dbg> zmk.hid_listener_keycode_pressed: usage_page 0x07 keycode 0xe2 implicit_mods 0x00 explicit_mods 0x00
[00:32:16.678,314] <dbg> zmk.zmk_hid_register_mod: Modifier 2 count 1
[00:32:16.678,314] <dbg> zmk.zmk_hid_register_mod: Modifiers set to 0x0d
[00:32:16.678,344] <dbg> zmk.zmk_hid_implicit_modifiers_press: Modifiers set to 0x0d
[00:32:16.678,344] <dbg> zmk.zmk_endpoints_send_report: usage page 0x07
[00:32:16.678,405] <dbg> zmk.release_captured_events: Releasing key position event for position 13 pressed
[00:32:16.678,405] <dbg> zmk.position_state_changed_listener: 13 bubble (no undecided hold_tap active)
[00:32:16.678,436] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:32:16.678,436] <dbg> zmk.release_captured_events: Releasing key position event for position 13 released
[00:32:16.678,466] <dbg> zmk.position_state_changed_listener: 13 bubble (no undecided hold_tap active)
[00:32:16.678,466] <dbg> zmk.zmk_keymap_apply_position_state: layer: 0 position: 13, binding name: MOD_TAP
[00:32:16.678,497] <dbg> zmk.on_hold_tap_binding_pressed: 13 new undecided hold_tap
[00:32:16.678,527] <dbg> zmk.zmk_keymap_apply_position_state: layer: 0 position: 13, binding name: MOD_TAP
[00:32:16.678,588] <dbg> zmk.decide_hold_tap: 13 decided hold (tap-preferred event 3)
[00:32:16.678,619] <dbg> zmk.on_keymap_binding_pressed: position 13 keycode 0x700e1
[00:32:16.678,649] <dbg> zmk.hid_listener_keycode_pressed: usage_page 0x07 keycode 0xe1 implicit_mods 0x00 explicit_mods 0x00
[00:32:16.678,649] <dbg> zmk.zmk_hid_register_mod: Modifier 1 count 1
[00:32:16.678,649] <dbg> zmk.zmk_hid_register_mod: Modifiers set to 0x0f
[00:32:16.678,680] <dbg> zmk.zmk_hid_implicit_modifiers_press: Modifiers set to 0x0f
[00:32:16.678,680] <dbg> zmk.zmk_endpoints_send_report: usage page 0x07
[00:32:16.686,309] <dbg> zmk.on_keymap_binding_released: position 13 keycode 0x700e1
[00:32:16.686,340] <dbg> zmk.hid_listener_keycode_released: usage_page 0x07 keycode 0xe1 implicit_mods 0x00 explicit_mods 0x00
[00:32:16.686,370] <dbg> zmk.zmk_hid_unregister_mod: Modifier 1 count: 0
[00:32:16.686,370] <dbg> zmk.zmk_hid_unregister_mod: Modifier 1 released
[00:32:16.686,370] <dbg> zmk.zmk_hid_unregister_mod: Modifiers set to 0x0d
[00:32:16.686,370] <dbg> zmk.zmk_hid_implicit_modifiers_release: Modifiers set to 0x0d
[00:32:16.686,370] <dbg> zmk.zmk_endpoints_send_report: usage page 0x07
[00:32:16.694,213] <dbg> zmk.on_hold_tap_binding_released: 13 cleaning up hold-tap
[00:32:16.694,244] <dbg> zmk.position_state_changed_listener: 536871936 bubbling 10 up event
[00:32:16.694,274] <dbg> zmk.zmk_keymap_apply_position_state: layer: 0 position: 10, binding name: MOD_TAP
[00:32:16.694,305] <dbg> zmk.on_keymap_binding_released: position 10 keycode 0x700e0
[00:32:16.694,335] <dbg> zmk.hid_listener_keycode_released: usage_page 0x07 keycode 0xe0 implicit_mods 0x00 explicit_mods 0x00
[00:32:16.694,335] <dbg> zmk.zmk_hid_unregister_mod: Modifier 0 count: 0
[00:32:16.694,366] <dbg> zmk.zmk_hid_unregister_mod: Modifier 0 released
[00:32:16.694,366] <dbg> zmk.zmk_hid_unregister_mod: Modifiers set to 0x0c
[00:32:16.694,366] <dbg> zmk.zmk_hid_implicit_modifiers_release: Modifiers set to 0x0c
[00:32:16.694,366] <dbg> zmk.zmk_endpoints_send_report: usage page 0x07
[00:32:16.702,209] <dbg> zmk.on_hold_tap_binding_released: 10 cleaning up hold-tap
[00:32:16.702,239] <dbg> zmk.zmk_kscan_process_msgq: Row: 1, col: 2, position: 11, pressed: false
[00:32:16.702,270] <dbg> zmk.position_state_changed_listener: 11 bubble (no undecided hold_tap active)
[00:32:16.702,301] <dbg> zmk.zmk_keymap_apply_position_state: layer: 0 position: 11, binding name: MOD_TAP
[00:32:16.702,331] <dbg> zmk.on_keymap_binding_released: position 11 keycode 0x700e3
[00:32:16.702,362] <dbg> zmk.hid_listener_keycode_released: usage_page 0x07 keycode 0xe3 implicit_mods 0x00 explicit_mods 0x00
[00:32:16.702,392] <dbg> zmk.zmk_hid_unregister_mod: Modifier 3 count: 0
[00:32:16.702,392] <dbg> zmk.zmk_hid_unregister_mod: Modifier 3 released
[00:32:16.702,392] <dbg> zmk.zmk_hid_unregister_mod: Modifiers set to 0x04
[00:32:16.702,392] <dbg> zmk.zmk_hid_implicit_modifiers_release: Modifiers set to 0x04
[00:32:16.702,392] <dbg> zmk.zmk_endpoints_send_report: usage page 0x07
[00:32:16.710,235] <dbg> zmk.on_hold_tap_binding_released: 11 cleaning up hold-tap
[00:32:16.710,266] <dbg> zmk.zmk_kscan_process_msgq: Row: 1, col: 3, position: 12, pressed: false
[00:32:16.710,296] <dbg> zmk.position_state_changed_listener: 12 bubble (no undecided hold_tap active)
[00:32:16.710,296] <dbg> zmk.zmk_keymap_apply_position_state: layer: 0 position: 12, binding name: MOD_TAP
[00:32:16.710,357] <dbg> zmk.on_keymap_binding_released: position 12 keycode 0x700e2
[00:32:16.710,388] <dbg> zmk.hid_listener_keycode_released: usage_page 0x07 keycode 0xe2 implicit_mods 0x00 explicit_mods 0x00
[00:32:16.710,388] <dbg> zmk.zmk_hid_unregister_mod: Modifier 2 count: 0
[00:32:16.710,388] <dbg> zmk.zmk_hid_unregister_mod: Modifier 2 released
[00:32:16.710,388] <dbg> zmk.zmk_hid_unregister_mod: Modifiers set to 0x00
[00:32:16.710,418] <dbg> zmk.zmk_hid_implicit_modifiers_release: Modifiers set to 0x00
[00:32:16.710,418] <dbg> zmk.zmk_endpoints_send_report: usage page 0x07
[00:32:16.718,261] <dbg> zmk.on_hold_tap_binding_released: 12 cleaning up hold-tap
```

**Comment by okke-formsma** [comment]
What I do not understand from the above logs are the following lines:
```
[00:32:10.889,831] <dbg> zmk.position_state_changed_listener: 11 bubble (no undecided hold_tap active)
[00:32:10.889,862] <dbg> zmk.zmk_keymap_apply_position_state: layer: 0 position: 10, binding name: MOD_TAP
[00:32:10.889,923] <dbg> zmk.on_hold_tap_binding_pressed: 10 new undecided hold_tap
[00:32:10.889,953] <dbg> zmk.position_state_changed_listener: 10 capturing 11 down event
```
position_state_changed_listener is somehow called twice for position event 11... :thinking:

**Comment by okke-formsma** [comment]
the position_state_changed is correctly triggered twice, because the combo retriggers it.

there seems to be some kind of timing issue with the combos though:

```
[00:00:23.139,007] <dbg> zmk.kscan_gpio_read_0: Sending event at 1,2 state on
[00:00:23.139,007] <dbg> zmk.kscan_gpio_read_0: Sending event at 1,4 state on
[00:00:23.139,068] <dbg> zmk.zmk_kscan_process_msgq: Row: 1, col: 2, position: 11, pressed: true
[00:00:23.139,099] <dbg> zmk.position_state_changed_listener: bubble pos 11 (no undecided hold_tap active)
[00:00:23.139,129] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:00:23.139,129] <dbg> zmk.zmk_kscan_process_msgq: Row: 1, col: 4, position: 13, pressed: true
[00:00:23.139,160] <dbg> zmk.position_state_changed_listener: bubble pos 13 (no undecided hold_tap active)
[00:00:23.139,190] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:00:23.210,815] <dbg> zmk.kscan_gpio_read_0: Sending event at 1,4 state off
[00:00:23.210,876] <dbg> zmk.zmk_kscan_process_msgq: Row: 1, col: 4, position: 13, pressed: false
[00:00:23.210,906] <dbg> zmk.position_state_changed_listener: bubble pos 13 (no undecided hold_tap active)
[00:00:23.210,906] <dbg> zmk.release_pressed_keys: Combo: release first captured event and reraise the rest.
[00:00:23.210,937] <dbg> zmk.zmk_keymap_apply_position_state: layer: 0 position: 11, binding name: MOD_TAP
[00:00:23.210,968] <dbg> zmk.on_hold_tap_binding_pressed: 11 new undecided hold_tap
[00:00:23.210,968] <dbg> zmk.on_hold_tap_binding_pressed: 79 tapping term left
[00:00:23.210,998] <dbg> zmk.position_state_changed_listener: ev timestamp 4294990585 sum 257698037761
[00:00:23.211,029] <dbg> zmk.position_state_changed_listener: 11 capturing pos 13 down event
[00:00:23.211,029] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:00:23.211,029] <dbg> zmk.zmk_keymap_apply_position_state: layer: 0 position: 13, binding name: MOD_TAP
[00:00:23.211,059] <err> zmk: ACTIVE_HOLD_TAP_CLEANED_UP_TOO_EARLY
```

**Comment by okke-formsma** [comment]
fixed in #694
